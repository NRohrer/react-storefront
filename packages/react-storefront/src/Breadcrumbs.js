/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import ArrowRight from '@material-ui/icons/KeyboardArrowRight'
import Link from './Link'
import withStyles from '@material-ui/core/styles/withStyles'
import Typography from '@material-ui/core/Typography'
import { inject, observer } from 'mobx-react'
import Track from './Track'
import Container from './Container'
import PropTypes from 'prop-types'

export const styles = theme => ({
  breadcrumbs: {
    margin: '0',
    backgroundColor: '#F4F2F1',
    padding: '12px 0',

    '& a': {
      color: theme.palette.text.primary,
      textDecoration: 'none'
    }
  },

  separator: {
    height: '12px',
    position: 'relative',
    top: '2px',
    width: '16px'
  },

  current: {
    fontWeight: 'bold',
    color: theme.palette.text.primary
  }
})

/**
 * Renders a list of breadcrumbs from AppModelBase.breadcrumbs.  When a breadcrumb is clicked, the breadcrumb_clicked
 * analytics event is automatically fired.
 */
@withStyles(styles, { name: 'RSFBreadcrumbs' })
@inject('app')
@observer
export default class Breadcrumbs extends Component {
  static propTypes = {
    /**
     * The items to display, each with url, text, and state.
     */
    items: PropTypes.arrayOf(PropTypes.objectOf),

    /**
     * Set to true to hide the last breadcrumb
     */
    hideLast: PropTypes.bool
  }

  static defaultProps = {
    hideLast: false
  }

  render() {
    let {
      app: { breadcrumbs },
      items,
      hideLast,
      classes
    } = this.props

    breadcrumbs = items || breadcrumbs

    if (hideLast) {
      breadcrumbs = breadcrumbs.slice(0, breadcrumbs.length - 1)
    }

    return (
      <Typography className={classes.breadcrumbs} variant="caption">
        <Container>
          {breadcrumbs && breadcrumbs.map(this.renderBreadcrumb)}
          <span>&nbsp;</span>
        </Container>
      </Typography>
    )
  }

  renderBreadcrumb = (item, i) => {
    const { classes } = this.props
    const arrow = i > 0 ? <ArrowRight className={classes.separator} /> : null

    if (item.url) {
      return (
        <span key={i}>
          {arrow}
          <Track event="breadcrumbClicked" breadcrumb={item}>
            <Link to={item.url} state={item.state}>
              {item.text}
            </Link>
          </Track>
        </span>
      )
    } else {
      return (
        <span key={i} className={classes.current}>
          {arrow}
          {item.text}
        </span>
      )
    }
  }
}
