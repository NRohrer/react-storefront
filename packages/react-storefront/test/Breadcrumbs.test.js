/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'mobx-react'
import Breadcrumbs from '../src/Breadcrumbs'
import AppModelBase from '../src/model/AppModelBase'
import waitForAnalytics from './helpers/waitForAnalytics'
import TestProvider from './TestProvider'
import AnalyticsProvider from '../src/AnalyticsProvider'

describe('Breadcrumbs', () => {
  it('renders', () => {
    const app = AppModelBase.create({
      breadcrumbs: [
        { url: '/', text: 'Home' },
        { url: '/c/1', text: 'Category 1' },
        { text: 'Subcategory 1' }
      ]
    })
    expect(
      mount(
        <Provider app={app}>
          <Breadcrumbs />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('renders blank when no breadcrumbs are provided', () => {
    const app = AppModelBase.create({})
    expect(
      mount(
        <Provider app={app}>
          <Breadcrumbs />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('renders blank when an empty array of breadcrumbs is provided', () => {
    const app = AppModelBase.create({ breadcrumbs: [] })
    expect(
      mount(
        <Provider app={app}>
          <Breadcrumbs />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('includes state in the link', () => {
    const app = AppModelBase.create({
      breadcrumbs: [
        { url: '/', text: 'Home' },
        { url: '/c/1', text: 'Category 1', state: { loadingCategory: { title: 'Category 1' } } },
        { text: 'Subcategory 1', state: { loadingSubcategory: { title: 'Subcategory 1' } } }
      ]
    })
    expect(
      mount(
        <Provider app={app}>
          <Breadcrumbs />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('should fire a breadcrumb_clicked event when clicked', () => {
    const breadcrumbClicked = jest.fn()

    const app = AppModelBase.create({
      breadcrumbs: [
        { url: '/', text: 'Home' },
        { url: '/c/1', text: 'Category 1' },
        { text: 'Subcategory 1' }
      ]
    })

    const wrapper = mount(
      <TestProvider app={app}>
        <AnalyticsProvider targets={() => [{ breadcrumbClicked }]}>
          <Breadcrumbs />
        </AnalyticsProvider>
      </TestProvider>
    )

    wrapper
      .find('a')
      .at(0)
      .simulate('click')

    return waitForAnalytics(() => {
      expect(breadcrumbClicked).toHaveBeenCalledWith({
        breadcrumb: {
          text: 'Home',
          url: '/',
          state: undefined
        }
      })
    })
  })

  describe('hideLast', () => {
    it('should not show the last breadcrumb when true', () => {
      const app = AppModelBase.create({
        breadcrumbs: [
          { url: '/', text: 'Home' },
          {
            url: '/c/1',
            text: 'Category 1',
            state: { loadingCategory: { title: 'Category 1' } }
          },
          { text: 'Subcategory 1', state: { loadingSubcategory: { title: 'Subcategory 1' } } }
        ]
      })

      const wrapper = mount(
        <Provider app={app}>
          <Breadcrumbs hideLast />
        </Provider>
      )

      expect(wrapper.findWhere(n => n.text() === 'Subcategory 1')).toHaveLength(0)
    })
    it('should show the last breadcrumb when false', () => {
      const app = AppModelBase.create({
        breadcrumbs: [
          { url: '/', text: 'Home' },
          {
            url: '/c/1',
            text: 'Category 1',
            state: { loadingCategory: { title: 'Category 1' } }
          },
          { text: 'Subcategory 1', state: { loadingSubcategory: { title: 'Subcategory 1' } } }
        ]
      })

      const wrapper = mount(
        <Provider app={app}>
          <Breadcrumbs />
        </Provider>
      )

      expect(wrapper.findWhere(n => n.text() === 'Subcategory 1').length).toBeGreaterThan(0)
    })
  })
})
