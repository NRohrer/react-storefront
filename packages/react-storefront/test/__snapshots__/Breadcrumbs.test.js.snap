// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs includes state in the link 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Breadcrumbs-with-app)>
    <inject-Breadcrumbs-with-app classes={{...}}>
      <Breadcrumbs classes={{...}} app={{...}} hideLast={false}>
        <WithStyles(Typography) className="RSFBreadcrumbs-breadcrumbs-1" variant="caption">
          <Typography className="RSFBreadcrumbs-breadcrumbs-1" variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
            <span className="MuiTypography-root-4 MuiTypography-caption-14 RSFBreadcrumbs-breadcrumbs-1">
              <WithStyles(Container)>
                <Container classes={{...}} inset={true}>
                  <div className="RSFContainer-root-39 RSFContainer-inset-40">
                    <span>
                      <inject-Track-with-app event="breadcrumbClicked" breadcrumb={{...}}>
                        <Track event="breadcrumbClicked" breadcrumb={{...}} app={{...}} trigger="onClick" onSuccess={[Function]} ampData={{...}}>
                          <WithStyles(inject-Link) to="/" state={[undefined]} onClick={[Function]}>
                            <inject-Link to="/" state={[undefined]} onClick={[Function]} classes={{...}}>
                              <Link to="/" state={[undefined]} onClick={[Function]} classes={{...}} history={[undefined]} router={[undefined]} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-41" style={[undefined]} onClick={[Function]} href="/">
                                  Home
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </Track>
                      </inject-Track-with-app>
                    </span>
                    <span>
                      <pure(KeyboardArrowRightIcon) className="RSFBreadcrumbs-separator-2">
                        <KeyboardArrowRightIcon className="RSFBreadcrumbs-separator-2">
                          <WithStyles(SvgIcon) className="RSFBreadcrumbs-separator-2">
                            <SvgIcon className="RSFBreadcrumbs-separator-2" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-42 RSFBreadcrumbs-separator-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowRightIcon>
                      </pure(KeyboardArrowRightIcon)>
                      <inject-Track-with-app event="breadcrumbClicked" breadcrumb={{...}}>
                        <Track event="breadcrumbClicked" breadcrumb={{...}} app={{...}} trigger="onClick" onSuccess={[Function]} ampData={{...}}>
                          <WithStyles(inject-Link) to="/c/1" state={{...}} onClick={[Function]}>
                            <inject-Link to="/c/1" state={{...}} onClick={[Function]} classes={{...}}>
                              <Link to="/c/1" state={{...}} onClick={[Function]} classes={{...}} history={[undefined]} router={[undefined]} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-41" style={[undefined]} onClick={[Function]} href="/c/1">
                                  Category 1
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </Track>
                      </inject-Track-with-app>
                    </span>
                    <span className="RSFBreadcrumbs-current-3">
                      <pure(KeyboardArrowRightIcon) className="RSFBreadcrumbs-separator-2">
                        <KeyboardArrowRightIcon className="RSFBreadcrumbs-separator-2">
                          <WithStyles(SvgIcon) className="RSFBreadcrumbs-separator-2">
                            <SvgIcon className="RSFBreadcrumbs-separator-2" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-42 RSFBreadcrumbs-separator-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowRightIcon>
                      </pure(KeyboardArrowRightIcon)>
                      Subcategory 1
                    </span>
                    <span>
                      Â 
                    </span>
                  </div>
                </Container>
              </WithStyles(Container)>
            </span>
          </Typography>
        </WithStyles(Typography)>
      </Breadcrumbs>
    </inject-Breadcrumbs-with-app>
  </WithStyles(inject-Breadcrumbs-with-app)>
</Provider>
`;

exports[`Breadcrumbs renders 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Breadcrumbs-with-app)>
    <inject-Breadcrumbs-with-app classes={{...}}>
      <Breadcrumbs classes={{...}} app={{...}} hideLast={false}>
        <WithStyles(Typography) className="RSFBreadcrumbs-breadcrumbs-1" variant="caption">
          <Typography className="RSFBreadcrumbs-breadcrumbs-1" variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
            <span className="MuiTypography-root-4 MuiTypography-caption-14 RSFBreadcrumbs-breadcrumbs-1">
              <WithStyles(Container)>
                <Container classes={{...}} inset={true}>
                  <div className="RSFContainer-root-39 RSFContainer-inset-40">
                    <span>
                      <inject-Track-with-app event="breadcrumbClicked" breadcrumb={{...}}>
                        <Track event="breadcrumbClicked" breadcrumb={{...}} app={{...}} trigger="onClick" onSuccess={[Function]} ampData={{...}}>
                          <WithStyles(inject-Link) to="/" state={[undefined]} onClick={[Function]}>
                            <inject-Link to="/" state={[undefined]} onClick={[Function]} classes={{...}}>
                              <Link to="/" state={[undefined]} onClick={[Function]} classes={{...}} history={[undefined]} router={[undefined]} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-41" style={[undefined]} onClick={[Function]} href="/">
                                  Home
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </Track>
                      </inject-Track-with-app>
                    </span>
                    <span>
                      <pure(KeyboardArrowRightIcon) className="RSFBreadcrumbs-separator-2">
                        <KeyboardArrowRightIcon className="RSFBreadcrumbs-separator-2">
                          <WithStyles(SvgIcon) className="RSFBreadcrumbs-separator-2">
                            <SvgIcon className="RSFBreadcrumbs-separator-2" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-42 RSFBreadcrumbs-separator-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowRightIcon>
                      </pure(KeyboardArrowRightIcon)>
                      <inject-Track-with-app event="breadcrumbClicked" breadcrumb={{...}}>
                        <Track event="breadcrumbClicked" breadcrumb={{...}} app={{...}} trigger="onClick" onSuccess={[Function]} ampData={{...}}>
                          <WithStyles(inject-Link) to="/c/1" state={[undefined]} onClick={[Function]}>
                            <inject-Link to="/c/1" state={[undefined]} onClick={[Function]} classes={{...}}>
                              <Link to="/c/1" state={[undefined]} onClick={[Function]} classes={{...}} history={[undefined]} router={[undefined]} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-41" style={[undefined]} onClick={[Function]} href="/c/1">
                                  Category 1
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </Track>
                      </inject-Track-with-app>
                    </span>
                    <span className="RSFBreadcrumbs-current-3">
                      <pure(KeyboardArrowRightIcon) className="RSFBreadcrumbs-separator-2">
                        <KeyboardArrowRightIcon className="RSFBreadcrumbs-separator-2">
                          <WithStyles(SvgIcon) className="RSFBreadcrumbs-separator-2">
                            <SvgIcon className="RSFBreadcrumbs-separator-2" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                              <svg className="MuiSvgIcon-root-42 RSFBreadcrumbs-separator-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
                                <path fill="none" d="M0 0h24v24H0V0z" />
                              </svg>
                            </SvgIcon>
                          </WithStyles(SvgIcon)>
                        </KeyboardArrowRightIcon>
                      </pure(KeyboardArrowRightIcon)>
                      Subcategory 1
                    </span>
                    <span>
                      Â 
                    </span>
                  </div>
                </Container>
              </WithStyles(Container)>
            </span>
          </Typography>
        </WithStyles(Typography)>
      </Breadcrumbs>
    </inject-Breadcrumbs-with-app>
  </WithStyles(inject-Breadcrumbs-with-app)>
</Provider>
`;

exports[`Breadcrumbs renders blank when an empty array of breadcrumbs is provided 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Breadcrumbs-with-app)>
    <inject-Breadcrumbs-with-app classes={{...}}>
      <Breadcrumbs classes={{...}} app={{...}} hideLast={false}>
        <WithStyles(Typography) className="RSFBreadcrumbs-breadcrumbs-1" variant="caption">
          <Typography className="RSFBreadcrumbs-breadcrumbs-1" variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
            <span className="MuiTypography-root-4 MuiTypography-caption-14 RSFBreadcrumbs-breadcrumbs-1">
              <WithStyles(Container)>
                <Container classes={{...}} inset={true}>
                  <div className="RSFContainer-root-39 RSFContainer-inset-40">
                    <span>
                      Â 
                    </span>
                  </div>
                </Container>
              </WithStyles(Container)>
            </span>
          </Typography>
        </WithStyles(Typography)>
      </Breadcrumbs>
    </inject-Breadcrumbs-with-app>
  </WithStyles(inject-Breadcrumbs-with-app)>
</Provider>
`;

exports[`Breadcrumbs renders blank when no breadcrumbs are provided 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Breadcrumbs-with-app)>
    <inject-Breadcrumbs-with-app classes={{...}}>
      <Breadcrumbs classes={{...}} app={{...}} hideLast={false}>
        <WithStyles(Typography) className="RSFBreadcrumbs-breadcrumbs-1" variant="caption">
          <Typography className="RSFBreadcrumbs-breadcrumbs-1" variant="caption" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
            <span className="MuiTypography-root-4 MuiTypography-caption-14 RSFBreadcrumbs-breadcrumbs-1">
              <WithStyles(Container)>
                <Container classes={{...}} inset={true}>
                  <div className="RSFContainer-root-39 RSFContainer-inset-40">
                    <span>
                      Â 
                    </span>
                  </div>
                </Container>
              </WithStyles(Container)>
            </span>
          </Typography>
        </WithStyles(Typography)>
      </Breadcrumbs>
    </inject-Breadcrumbs-with-app>
  </WithStyles(inject-Breadcrumbs-with-app)>
</Provider>
`;
